firedrake/adjoint_utils/blocks/function.py

```diff
@@ -109,14 +107,14 @@ def _adj_assign_constant(self, adj_output, constant_fs):
         shape = r.ufl_shape
         if shape == () or shape[0] == 1:
             # Scalar Constant
-            r.vector()[:] = adj_output.vector().sum()
+            r.dat.data[:] = adj_output.dat.data_ro.sum()
```

Comment:
Nitpicking, but this I think _should_ be `r.dat.data_ro[:]`

Reply:
```suggestion
            r.dat.data_ro[:] = adj_output.dat.data_ro.sum()
```

Reply:
We are modifying the value of `r` so why should we use `dat.data_ro` ? `.dat.data` needs to be used whenever you change the value of the `.dat`, right ?

Reply:
Apologies for the confusion, I meant to say `dat.data_wo` and hastily edited the comment after making it but clearly it didn't make it to you in time! I think `dat.data_wo` is correct here since you don't read the value

Reply:
Well, `Dat.data_wo` is the same as `Dat.data` (see https://github.com/OP2/PyOP2/blame/master/pyop2/types/dat.py#L223). Also, we abusively use `Dat.data` when changing the value of the `Dat` all over the code so I don't think that is a problem.